src/Microsoft.DotNet.Interactive/Commands/KernelCommand.cs

```diff
@@ -5,60 +5,81 @@
 using System.CommandLine.Parsing;
 using System.Diagnostics;
+using System.Linq;
 using System.Threading.Tasks;
 using System.Text.Json.Serialization;
 
-namespace Microsoft.DotNet.Interactive.Commands
+namespace Microsoft.DotNet.Interactive.Commands;
+
+[DebuggerStepThrough]
+public abstract class KernelCommand
 {
-    [DebuggerStepThrough]
-    public abstract class KernelCommand
+    private readonly RoutingSlip _routingSlip;
+
+    protected KernelCommand(
+        string targetKernelName = null, 
+        KernelCommand parent = null)
     {
-        protected KernelCommand(
-            string targetKernelName = null, 
-            KernelCommand parent = null)
-        {
-            Properties = new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase);
-            TargetKernelName = targetKernelName;
-            Parent = parent;
-            RoutingSlip = new RoutingSlip();
-        }
+        Properties = new Dictionary<string, object>(StringComparer.InvariantCultureIgnoreCase);
+        TargetKernelName = targetKernelName;
+        Parent = parent;
+        _routingSlip = new RoutingSlip();
+    }
 
-        [JsonIgnore] 
-        public KernelCommandInvocation Handler { get; set; }
+    [JsonIgnore] 
+    public KernelCommandInvocation Handler { get; set; }
 
-        [JsonIgnore]
-        public KernelCommand Parent { get; internal set; }
+    [JsonIgnore]
+    public KernelCommand Parent { get; internal set; }
 
-        [JsonIgnore]
-        public IDictionary<string, object> Properties { get; }
+    [JsonIgnore]
+    public IDictionary<string, object> Properties { get; }
 
-        public string TargetKernelName { get; internal set; }
+    public string TargetKernelName { get; internal set; }
 
-        internal static KernelCommand None => new NoCommand();
+    internal static KernelCommand None => new NoCommand();
 
-        public Uri OriginUri { get; set; }
+    public Uri OriginUri { get; set; }
 
-        public Uri DestinationUri { get; set; }
+    public Uri DestinationUri { get; set; }
 
-        [JsonIgnore]
-        internal SchedulingScope SchedulingScope { get; set; } 
+    [JsonIgnore]
+    internal SchedulingScope SchedulingScope { get; set; } 
 
-        [JsonIgnore]
-        internal bool? ShouldPublishCompletionEvent { get; set; }
+    [JsonIgnore]
+    internal bool? ShouldPublishCompletionEvent { get; set; }
 
-        [JsonIgnore] 
-        public ParseResult KernelChooserParseResult { get; internal set; }
+    [JsonIgnore] 
+    public ParseResult KernelChooserParseResult { get; internal set; }
 
-        [JsonIgnore] public RoutingSlip RoutingSlip { get; }
+    [JsonIgnore]
+    public IReadOnlyCollection<Uri> RoutingSlip
+    {
+        get
+        {
+            var unique = new HashSet<Uri>();
+            var list = new List<Uri>(Parent?.RoutingSlip ?? Array.Empty<Uri>());
+            list.AddRange(_routingSlip.Where(unique.Add));
+            return list;
```

Comment:
Properties shouldn't return a new instance on each call. Is there a reason why this state isn't being built up only once?

Reply:
changed that to be built when parenting is changed